From 6d170fca988478ac4ee5cb8309dbc58433476abd Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Thu, 18 Apr 2024 10:53:19 +0200
Subject: [PATCH 1119/1141] drm/stm: ltdc: set default parent of pixel clock

Parent of the pixel clock should default to the reference clock
(rcc clock). If the driver has already been started, this action
is not necessary and may cause an issue on register reading/writing.

Change-Id: Ic9c1227260acaf33c70df07d4a4c3b6228b0a51e
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/375589
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/gpu/drm/stm/ltdc.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

--- a/drivers/gpu/drm/stm/ltdc.c
+++ b/drivers/gpu/drm/stm/ltdc.c
@@ -2601,6 +2601,7 @@ int ltdc_parse_device_tree(struct device
 int ltdc_get_clk(struct device *dev, struct ltdc_device *ldev)
 {
 	struct device_node *node;
+	int ret;
 
 	DRM_DEBUG_DRIVER("\n");
 
@@ -2638,6 +2639,18 @@ int ltdc_get_clk(struct device *dev, str
 			}
 			of_node_put(node);
 		}
+
+		/*
+		 * Parent of the pixel clock should default to the reference clock (rcc clock).
+		 * If the driver has already been started, this action is not necessary and
+		 *  may cause an issue on register reading/writing.
+		 */
+		if (!device_property_read_bool(dev, "default-on")) {
+			ret = clk_set_parent(ldev->pixel_clk, ldev->ltdc_clk);
+			if (ret)
+				return dev_err_probe(dev, PTR_ERR(ldev->lvds_clk),
+						     "Could not set parent clock\n");
+		}
 	}
 
 	return 0;
